File: Capitulo3-Interpolacao/lagrange.m
%****************************************************************************80
%  Code: 
%   lagrange.m 
%
%  Purpose:
%   Implements the Lagrange Interpolation Px
%   
%  Modified:
%   Jul 25 2021 16:34 
%
%  Author:
%    Murilo Do Carmo Boratto [muriloboratto@uneb.br]  
%   
%  How to Execute:
%   [C]   = lagrange(X,Y)
%   [C,L] = lagrange(X,Y)
%  
%  Comments:
%
%    Input parameters:
%      X abscissa vector
%      Y ordinate vector
%
%    Output parameters:
%      C vector with the coefficients of the Lagrange interpolator polynomial
%      L matrix with the coefficients of the Lagrange polynomials
%*****************************************************************************

function [C,L] = lagrange(X,Y)
echo off;

n = length(X);             
L = zeros(n,n);            

for k=1:n,               
  V = 1;
  for i=1:n,
    if i ~= k
      V = conv(V,poly(X(i)));
      V = V/(X(k)-X(i)); 
    end
  end

  L(k,:) = V;
   
end

C = Y * L;                   
                                                     

File: Capitulo3-Interpolacao/lagrange.py
# -*- coding: utf-8 -*-
#****************************************************************************80
#  Code: 
#   lagrange.py 
#
#  Purpose:
#   Implement the Lagrange Method in Python
#
#  Modified:
#   Aug 24 2021 16:34 
#
#  Author:
#    Murilo Do Carmo Boratto < muriloboratto 'at' uneb.br >  
#   
#  How to Execute:
#    python lagrange.py 
#
#*****************************************************************************

import numpy as np
from numpy.polynomial.polynomial import Polynomial
import pylab as plt
from scipy.interpolate import lagrange

x = [0, 1, 3]
y = [-5, 1, 25]

f = lagrange(x, y)

print(Polynomial(f).coef)

x_new = np.arange(0, 3.1, 0.1)

fig = plt.figure(figsize = (10,8))
plt.plot(x_new, f(x_new), 'b', x, y, 'ro')
plt.title('Lagrange Polynomial')
plt.grid()
plt.xlabel('x')
plt.ylabel('y')
plt.show()
